### ISSUE TITLE
Exa not moving (But single monitor works)

### ISSUE THREAD

**Issue by Aratietur** [MestreLion/hackmatch#4]
Hi i'm already working on the next issues :-D

When i'm only using my 2nd monitor the following happens, but i still need to switch to windowed and back to fullscreen:
```
C:\Users\nikol\AppData\Roaming\Python\Python311\Scripts>hackmatch-bot
Read game settings: C:\Users\nikol\OneDrive\Dokumente\My Games\EXAPUNKS\*\config.cfg
Launching game and waiting 60 seconds for game window
Launch game by Steam URI: steam://rungameid/716490
Read game settings: C:\Users\nikol\OneDrive\Dokumente\My Games\EXAPUNKS\*\config.cfg
Game window: <Win32Window left="0", top="0", width="1606", height="929", title="EXAPUNKS">
[2023-09-07 11:51:49 WARNIN] gui : Unsupported game window size: (1606, 929), must be one of ((1920, 1080), (1920, 1200), (1600, 900), (1366, 768))
Game window resized: (1920, 1080)
ygb..by
g.yyygy
..Gb...
.......
.......
.......
.......
.......
.......
___@___
MATCH FOUND! After 51ms, 704 boards (13751 boards/s), 6 moves deep
        SWAP, GRAB, RIGHT, RIGHT, GRAB, SWAP, LEFT, LEFT
```

And a lot more of these (in --verbose there are a lot of screenshots, can post the lines when needed).
But the Exa in Hackmatch didn't move at all...

I controlled the ingame config and CRT is disabled, and keymapping seems to be fine ( i can play with j&k)

**Comment by Aratietur** [comment]
Here you get a full log from start until loose:
- Using only 2nd monitor
- Fullscreen 1900*1200
- Window ingame 1600*900
- i started the game, moved to options switched fullscreen-->windowed-->fullscreen
- after that i started hackmatch as fast as possible
- i let the game run without doing anything until the match got lost

[Log (2).txt](https://github.com/MestreLion/hackmatch/files/12547636/Log.2.txt)

**Comment by MestreLion** [comment]
I was suspecting this when I skimmed over your first logs posted in the previous issues: even when the bot complains about "wrong window size" of 1606x929 instead of 1600x900, it **was** able to parse the game window image into a logical board, find matches and calculate the set of moves, even in your multi-monitor setup. So the "wrong" window size is *not* the actual issue here.

Btw, a side note: I believe this 1606x929 is due to the window borders... +6 pixels on width and +29 on height sounds consistent with the (thin) lateral borders and (thick) title bar. Usually `pywinctl` can tell apart such windows _decorations_ from its _content_,  and my bot accounts for that, but it seems that on Windows the full size is being considered as content, hence the warning. When on `--debug`, what are the dumped screenshots dimensions? Do the block overlays properly match the visual content?

So, back to the point: the issue here are the _*controls*_. Keycodes might be different from Linux to Windows (I assumed the game used SDL keyscan codes in both platforms).

The bot uses `PyAutoGUI` for sending the key presses to the game window, so we're looking at a different library now. While I investigate the issue, one helpful test you could make is to install the `[extra]` components, which currently is simply `pygame`. If `pygame` is installed, the bot uses it to augment the translation from game config file keycodes to PyAutoGUI keycodes. In other words: it doesn't (or shouldn't) matter how you configured your key mapping in-game, the bot reads that configuration and translates that to the format PyAutoGUI expects. So `pip install pygame` and see if there is a difference/improvement.

Another side-note: the usual workflow for the bot is:
- Use a single monitor, with a desktop resolution such as 1920x1080 or 1920x1200
- Set the game to windowed, with a dimension that comfortably fits the desktop  (such as 1600x900), **OR** set it to fullscreen matching your desktop resolution. The bot was extensively tested in 1920x1200, 1920x1080 and 1600x900. Not so much in 1366x768, tho. It works, but accuracy is low.
- I prefer windowed so I can align the game window and the terminal window running the bot  without overlapping the board area (and also manually set the terminal window to "always on top")
- Once the game is launched and its window is detected by the bot (within 60 seconds of launch), take all the time you want. The bot stays kind of idle until it detects the a hackmatch _board_.
- The board image parsing and AI only requires an **unobstructed** _board area_. The game window does not need to be focused or active. The bot movement input, however, does require the game window to be focused. Neither care about the "leaderboard" area being obstructed or not (as seen on my video).
- So one way to "pause" the bot is to either ALT+TAB switch to the bot terminal window (Board parsing and AI will keep working, but as input will go the the bot window itself instead of game window, the player won't move), or to slightly obstruct the board area (possibly with the bot window itself, or moving game window slightly off-screen) to the point the board parsing will stop working

**Comment by MestreLion** [comment]
My point with that last side-note is:
- No need to toggle to/from full-screen mid-game. It's better if the window is not resized at all
- No need to start hackmatch quickly after launching exapunks. As long as the bot detects the exapunks window within 60s after launch, it will work whenever it detects a hackmatch board area and be idle otherwise

**Comment by MestreLion** [comment]
Anyway, forget everything I said: yes, it _might_ have a platform issue with controls, but the debug log clearly shows it is not parsing the board properly. There are many "holes" in the parsed board, so I'm sure it does not represent the actual images. This happens in some frames on 1366x768, that's why I'm familiar to those "holes" in the logs, and that's why 1366x768 is not accurate. But you had holes in _all_ frames, so it _never_ properly identified the board elements.
Without accurate image parsing, the AI (and the bot) is useless, even if we manage to fix the controls.

My image parsing routine is quite flexible on _position_ (and size, to an extent), but quite picky on _colors_: there's no fuzzy matching, the pixels must have the _exact_ color for a block to be identified. It seems there's something on Windows (or SDL?) that makes the colors slightly different than on Linux. Maybe a global shader, or video card driver setting, or a desktop color correction, who knows?

One solution would be color fuzzy matching. Fidel's solver implements that, but I didn't. For 2 reasons:
- I'm lazy and I didn't _need_ it, as the colors in my system matched _perfectly_ with Fidel's (and Laelaths')  constants in _all_ resolutions
- Fuzzy color matching will surely have a big performance impact. Being Python my bot is already much slower than Fidel's Rust, and fuzzy matching will make it worse.

And performance _is_ a bottleneck: In my fairly powerful Ryzen 2600 my bot evaluates around 8000 outcomes on average in the default 850ms, and most of the time that's not enough to find the move sequence that leads to matching blocks, so it just times out and play the "best" move it can hoping for a good outcome that leads to matching eventually. (as a reference, Laelath's uses 110ms as its timeout)

The math is: the less outcomes evaluated per play, the less likely it finds block-matching moves, so the sooner it loses the game. That's why I can only _consistently_ reach 80-120K, some luck is required for 150K+, and I've never reached 200K. Fidel's can get to 500K and Laelath's (which, not surprisingly, does no fuzzy matching either)  goes to _millions_ of points. 

My board parsing routine is quite fast (and the screenshot is a huge chunk of that), but any time spent on board parsing is less time on AI, potentially lowering the max score.

That said, I do not oppose fuzzy matching in my bot. It's not even hard to implement it (and Fidel Solver can be used as a reference), but:
- Currently I don't have the time or motivation to implement it myself. PR's are welcome tho
- It should be disabled by default, enabled with a `--fuzzy` argument (or on Windows platforms), and have none to negligible performance hit if disabled

**Comment by pnemes** [comment]
see #7

**Comment by MestreLion** [comment]
Fixed by #7
